File: cmux-tui/bindings/conformance/adapters/python/adapter.py
#!/usr/bin/env python3
"""Public resource conformance adapter for the dependency-free Python SDK."""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Mapping


ROOT = Path(__file__).resolve().parents[5]
sys.path.insert(0, str(ROOT / "cmux-tui" / "bindings" / "python"))

import cmux  # noqa: E402


def error_value(error: cmux.ResourceError) -> dict[str, Any]:
    return {
        "code": error.code,
        "message": error.message,
        "details": plain(error.details),
        "retryable": error.retryable,
    }


def plain(value: Any) -> Any:
    if isinstance(value, cmux.Document):
        return plain(value.fields)
    if isinstance(value, cmux.Cursor):
        return {
            "generation": value.generation,
            "revision": str(value.revision),
        }
    if isinstance(value, Mapping):
        return {str(key): plain(item) for key, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [plain(item) for item in value]
    if isinstance(value, int) and not isinstance(value, bool):
        return value
    if hasattr(value, "value") and isinstance(value.value, str):
        return value.value
    return value


def session(client: cmux.Client, constants: Mapping[str, str]):
    return client.session(cmux.SessionId(constants["session"]))


def workspace(client: cmux.Client, constants: Mapping[str, str]):
    return session(client, constants).workspace(
        cmux.WorkspaceId(constants["workspace"])
    )


def mutation_value(result: Any) -> dict[str, Any]:
    handle = result.value
    snapshot = handle.snapshot
    if snapshot is None:
        snapshot = handle.refresh()
    return {
        "workspace_id": str(snapshot.id),
        "name": snapshot.name,
        "generation": result.generation,
        "revision": str(result.revision),
        "replayed": result.replayed,
    }


def unknown_value(item: Any) -> tuple[str, dict[str, Any]]:
    value = getattr(item, "value", getattr(item, "item", None))
    if not isinstance(value, cmux.Unknown):
        raise AssertionError("session event was not the public Unknown variant")
    return value.kind, plain(value.raw)


def drain_end(stream: Any) -> str:
    try:
        while True:
            next(stream)
    except cmux.StreamError as error:
        return error.reason
    except StopIteration:
        end = stream.end
        return end.reason if end is not None else "completed"


def run(payload: Mapping[str, Any]) -> Any:
    operation = payload["op"]
    constants = payload["constants"]
    if operation == "redaction":
        secret = "provider://conformance-secret"
        token = "renderer-conformance-secret"
        specifier = cmux.ExternalMachineSpecifier(secret)
        grant = cmux.RendererGrant(
            token,
            endpoint="unix:///tmp/renderer",
            terminal_id=cmux.TerminalId(
                "term_66666666666666666666666666666666"
            ),
            rights=("render",),
            ttl_ms=1000,
        )
        return {
            "specifier_redacted": secret not in repr(specifier)
            and secret not in str(specifier),
            "renderer_token_redacted": token not in repr(grant)
            and token not in str(grant),
        }

    with cmux.Client(
        payload["socket_path"],
        timeout=15.0,
        random_hex_128=lambda: "a" * 32,
    ) as client:
        if operation == "read":
            document = session(client, constants).ping()
            fields = plain(document)
            return {
                "alive": fields["alive"],
                "cursor": fields["cursor"],
            }
        if operation == "mutation-replay":
            target = workspace(client, constants)
            options = {
                "idempotency_key": constants["idempotency_key"],
                "expected_revision": constants["revision"],
            }
            first = target.rename(constants["name"], **options)
            second = target.rename(constants["name"], **options)
            return {
                "first": mutation_value(first),
                "second": mutation_value(second),
            }
        if operation == "mutation-error":
            try:
                workspace(client, constants).rename(
                    constants["name"],
                    idempotency_key=constants["idempotency_key"],
                    expected_revision=constants["revision"],
                )
            except cmux.ResourceError as error:
                return error_value(error)
            raise AssertionError("mutation unexpectedly succeeded")
        if operation == "stream-unknown":
            stream = session(client, constants).events()
            item = next(stream)
            kind, raw = unknown_value(item)
            try:
                next(stream)
            except StopIteration:
                pass
            return {
                "sequence": str(item.sequence),
                "cursor": plain(item.cursor),
                "kind": kind,
                "raw": raw,
                "end": stream.end.reason if stream.end else "completed",
            }
        if operation == "stream-cancel":
            stream = session(client, constants).events()
            stream.cancel()
            stream.cancel()
            count = 0
            try:
                while True:
                    next(stream)
                    count += 1
            except StopIteration:
                pass
            return {
                "end": stream.end.reason if stream.end else "canceled",
                "items_after_cancel": count,
                "cancel_calls": 2,
            }
        if operation == "stream-overflow":
            first = session(client, constants).events()
            first_end = drain_end(first)
            second = session(client, constants).events()
            second_item = next(second)
            second_kind, _ = unknown_value(second_item)
            try:
                next(second)
            except StopIteration:
                pass
            control = plain(session(client, constants).ping())
            return {
                "first_end": first_end,
                "second_kind": second_kind,
                "control_alive": control["alive"],
            }
        if operation == "live-flow":
            current = client.session(cmux.Selector.current())
            pinged = bool(plain(current.ping())["alive"])
            name = payload["workspace_name"]
            created = current.create_workspace(
                cmux.CreateWorkspaceOptions(name=name, initial_content="empty"),
                idempotency_key="live-create",
            )
            created_workspace = created.value.workspace
            if created_workspace is None:
                raise AssertionError("workspace.create omitted workspace handle")
            created_id = created_workspace.id
            renamed = created_workspace.rename(
                name + "-renamed", idempotency_key="live-rename"
            )
            listed = any(
                item.id == created_id
                for item in current.list_workspaces()
            )
            created_workspace.close(idempotency_key="live-close")
            disappeared = all(
                item.id != created_id
                for item in current.list_workspaces()
            )
            return {
                "pinged": pinged,
                "created": created_id is not None,
                "renamed": renamed.value.snapshot.name == name + "-renamed",
                "listed": listed,
                "closed": True,
                "disappeared": disappeared,
            }
    raise AssertionError(f"unknown adapter operation {operation}")


def main() -> int:
    payload = json.loads(sys.stdin.readline())
    identifier = payload.get("id")
    try:
        value = run(payload)
        response = {
            "contract_version": 2,
            "id": identifier,
            "ok": True,
            "value": value,
        }
    except BaseException as error:
        response = {
            "contract_version": 2,
            "id": identifier,
            "ok": False,
            "error": {
                "kind": "adapter",
                "message": f"{type(error).__name__}: {error}",
            },
        }
    print(json.dumps(response, separators=(",", ":"), ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: cmux-tui/bindings/conformance/adapters/typescript/adapter.mjs
#!/usr/bin/env node
// Public resource conformance adapter for the TypeScript Node entrypoint.

import {
  ExternalMachineSpecifier,
  RendererGrant,
  ResourceError,
  StreamError,
  decimalString,
  selectCurrent,
  selectName,
  sessionId,
  terminalId,
  workspaceId,
} from "../../../typescript/dist/src/index.js";
import {
  Client as ResourceClient,
  WebSocketTransport,
} from "../../../typescript/dist/src/browser.js";
import { NodeClient } from "../../../typescript/dist/src/node.js";

function errorValue(error) {
  return {
    code: error.code,
    message: error.message,
    details: error.details,
    retryable: error.retryable,
  };
}

function getSession(client, constants) {
  return client.session(sessionId(constants.session));
}

function getWorkspace(client, constants) {
  return getSession(client, constants).workspace(workspaceId(constants.workspace));
}

function mutationValue(result) {
  const snapshot = result.value.snapshot;
  if (!snapshot) throw new Error("mutation result handle omitted its snapshot");
  return {
    workspace_id: snapshot.id,
    name: snapshot.name,
    generation: result.generation,
    revision: result.revision,
    replayed: result.replayed,
  };
}

function unknownValue(item) {
  const value = item.value;
  if (
    typeof value?.kind === "string"
    && value.kind !== "snapshot"
    && value.kind !== "delta"
    && value.raw
    && typeof value.raw === "object"
  ) {
    return { kind: value.kind, raw: value.raw };
  }
  throw new Error("session event was not the public Unknown variant");
}

async function drainEnd(stream) {
  try {
    for await (const _item of stream) {
      // Deliberately drain until the terminal envelope.
    }
    return stream.end?.reason ?? "completed";
  } catch (error) {
    if (error instanceof StreamError) return error.reason;
    throw error;
  }
}

function makeClient(payload) {
  if (payload.transport === "websocket") {
    if (typeof payload.websocket_url !== "string") {
      throw new TypeError("websocket_url must be a string");
    }
    if (typeof payload.websocket_token !== "string") {
      throw new TypeError("websocket_token must be a string");
    }
    return new ResourceClient({
      transport: new WebSocketTransport(payload.websocket_url, {
        authToken: payload.websocket_token,
      }),
      timeoutMs: 15_000,
      randomHex128: () => "a".repeat(32),
    });
  }
  if (payload.transport !== undefined && payload.transport !== "unix") {
    throw new TypeError(`unsupported transport ${String(payload.transport)}`);
  }
  return new NodeClient({
    socketPath: payload.socket_path,
    timeoutMs: 15_000,
    randomHex128: () => "a".repeat(32),
  });
}

function liveSession(client) {
  return client.session(selectCurrent());
}

async function workspaceRows(current) {
  const rows = new Map();
  for (const item of await current.listWorkspaces()) {
    const snapshot = item.snapshot ?? await item.refresh();
    rows.set(snapshot.id, snapshot.name);
  }
  return rows;
}

function candidateIds(error) {
  if (
    !error.details
    || typeof error.details !== "object"
    || !Array.isArray(error.details.candidates)
  ) {
    return [];
  }
  return error.details.candidates.filter(
    (candidate) => typeof candidate === "string",
  );
}

async function liveSetup(client, baseName, keyPrefix) {
  const current = liveSession(client);
  const pinged = Boolean((await current.ping()).alive);
  const stableCreated = await current.createWorkspace(
    { name: baseName, initialContent: "empty" },
    { idempotencyKey: `${keyPrefix}-stable-create` },
  );
  const stable = stableCreated.value.workspace;
  if (!stable?.id) throw new Error("workspace.create omitted stable workspace handle");
  const stableId = stable.id;
  const stableRenamedName = `${baseName}-renamed`;
  const renamed = await stable.rename(stableRenamedName, {
    idempotencyKey: `${keyPrefix}-stable-rename`,
  });

  const duplicateName = `${baseName}-duplicate`;
  const duplicateIds = [];
  for (const suffix of ["a", "b"]) {
    const created = await current.createWorkspace(
      { name: duplicateName, initialContent: "empty" },
      { idempotencyKey: `${keyPrefix}-duplicate-${suffix}` },
    );
    const duplicate = created.value.workspace;
    if (!duplicate?.id) {
      throw new Error("workspace.create omitted duplicate workspace handle");
    }
    duplicateIds.push(duplicate.id);
  }

  let ambiguityCode = "";
  let ambiguityCandidates = [];
  try {
    await current.workspace(selectName(duplicateName)).rename(
      `${baseName}-must-not-apply`,
      { idempotencyKey: `${keyPrefix}-ambiguous-rename` },
    );
    throw new Error("duplicate workspace selector unexpectedly mutated");
  } catch (error) {
    if (!(error instanceof ResourceError)) throw error;
    ambiguityCode = error.code;
    ambiguityCandidates = candidateIds(error);
  }
  const rows = await workspaceRows(current);
  return {
    pinged,
    stable_id: stableId,
    stable_renamed: renamed.value.snapshot?.name === stableRenamedName,
    duplicate_ids: duplicateIds,
    ambiguity_code: ambiguityCode,
    ambiguity_preserved_all_candidates:
      ambiguityCandidates.length === duplicateIds.length
      && duplicateIds.every((identifier) => ambiguityCandidates.includes(identifier)),
    no_mutation:
      duplicateIds.every((identifier) => rows.get(identifier) === duplicateName)
      && ![...rows.values()].includes(`${baseName}-must-not-apply`),
  };
}

async function liveRestart(
  client,
  baseName,
  keyPrefix,
  expectedStableId,
  expectedDuplicateIds,
) {
  const current = liveSession(client);
  const rows = await workspaceRows(current);
  const expectedIds = [expectedStableId, ...expectedDuplicateIds];
  const sameIds = expectedIds.every((identifier) => rows.has(identifier));
  const stableNamePreserved =
    rows.get(expectedStableId) === `${baseName}-renamed`;
  const duplicatesPreserved = expectedDuplicateIds.every(
    (identifier) => rows.get(identifier) === `${baseName}-duplicate`,
  );

  await current.workspace(workspaceId(expectedStableId)).close({
    idempotencyKey: `${keyPrefix}-close-stable`,
  });
  for (const [index, identifier] of expectedDuplicateIds.entries()) {
    await current.workspace(workspaceId(identifier)).close({
      idempotencyKey: `${keyPrefix}-close-${index === 0 ? "a" : "b"}`,
    });
  }
  const remaining = await workspaceRows(current);
  return {
    same_ids: sameIds,
    stable_name_preserved: stableNamePreserved,
    duplicates_preserved: duplicatesPreserved,
    closed: true,
    disappeared: expectedIds.every((identifier) => !remaining.has(identifier)),
  };
}

async function run(payload) {
  const constants = payload.constants;
  if (payload.op === "redaction") {
    const specifierSecret = "provider://conformance-secret";
    const rendererSecret = "renderer-conformance-secret";
    const specifier = new ExternalMachineSpecifier(specifierSecret);
    const renderer = new RendererGrant(
      rendererSecret,
      "unix:///tmp/renderer",
      terminalId("term_66666666666666666666666666666666"),
      ["render"],
      1000,
    );
    return {
      specifier_redacted:
        !specifier.toString().includes(specifierSecret)
        && !JSON.stringify(specifier).includes(specifierSecret),
      renderer_token_redacted:
        !renderer.toString().includes(rendererSecret)
        && !JSON.stringify(renderer).includes(rendererSecret),
    };
  }

  const client = makeClient(payload);
  try {
    if (payload.op === "read") {
      const result = await getSession(client, constants).ping();
      return { alive: result.alive, cursor: result.cursor };
    }
    if (payload.op === "mutation-replay") {
      const target = getWorkspace(client, constants);
      const options = {
        idempotencyKey: constants.idempotency_key,
        expectedRevision: decimalString(constants.revision),
      };
      const first = await target.rename(constants.name, options);
      const second = await target.rename(constants.name, options);
      return {
        first: mutationValue(first),
        second: mutationValue(second),
      };
    }
    if (payload.op === "mutation-error") {
      try {
        await getWorkspace(client, constants).rename(constants.name, {
          idempotencyKey: constants.idempotency_key,
          expectedRevision: decimalString(constants.revision),
        });
      } catch (error) {
        if (error instanceof ResourceError) return errorValue(error);
        throw error;
      }
      throw new Error("mutation unexpectedly succeeded");
    }
    if (payload.op === "stream-unknown") {
      const stream = await getSession(client, constants).events();
      const next = await stream.next();
      if (next.done) throw new Error("unknown stream ended before its item");
      const unknown = unknownValue(next.value);
      await stream.next();
      return {
        sequence: next.value.sequence,
        cursor: next.value.cursor,
        ...unknown,
        end: stream.end?.reason ?? "completed",
      };
    }
    if (payload.op === "stream-cancel") {
      const stream = await getSession(client, constants).events();
      await stream.cancel();
      await stream.cancel();
      let count = 0;
      for await (const _item of stream) count += 1;
      return {
        end: stream.end?.reason ?? "canceled",
        items_after_cancel: count,
        cancel_calls: 2,
      };
    }
    if (payload.op === "stream-overflow") {
      const first = await getSession(client, constants).events();
      const firstEnd = await drainEnd(first);
      const second = await getSession(client, constants).events();
      const next = await second.next();
      if (next.done) throw new Error("independent stream ended before its item");
      const secondUnknown = unknownValue(next.value);
      await second.next();
      const control = await getSession(client, constants).ping();
      return {
        first_end: firstEnd,
        second_kind: secondUnknown.kind,
        control_alive: control.alive,
      };
    }
    if (payload.op === "live-setup") {
      return await liveSetup(client, payload.workspace_name, payload.key_prefix);
    }
    if (payload.op === "live-restart") {
      if (
        !Array.isArray(payload.expected_duplicate_ids)
        || !payload.expected_duplicate_ids.every(
          (identifier) => typeof identifier === "string",
        )
      ) {
        throw new TypeError("expected_duplicate_ids must be a string array");
      }
      return await liveRestart(
        client,
        payload.workspace_name,
        payload.key_prefix,
        payload.expected_stable_id,
        payload.expected_duplicate_ids,
      );
    }
    throw new Error(`unknown adapter operation ${payload.op}`);
  } finally {
    client.close();
  }
}

const chunks = [];
for await (const chunk of process.stdin) chunks.push(chunk);
const payload = JSON.parse(chunks.join(""));
let response;
try {
  response = {
    contract_version: 2,
    id: payload.id,
    ok: true,
    value: await run(payload),
  };
} catch (error) {
  response = {
    contract_version: 2,
    id: payload.id,
    ok: false,
    error: {
      kind: "adapter",
      message: `${error?.name ?? "Error"}: ${error?.message ?? String(error)}`,
    },
  };
}
process.stdout.write(`${JSON.stringify(response)}\n`);
